### DISCUSSION TITLE
Convert the earlier Spacy Tuples to training data for version 3

### DISCUSSION THREAD

**Discussion by anselmn** [explosion/spaCy#7150] [Help: Coding & Implementations] [answered]
hi Team, I am new to spacy and just late last year got started on spacy version 2, I now want to adopt to the latest version 3, but I am clueless of how to build my training dataset. 
In version 2 I had created a set of Tuples of the form --- 
```
[("Hello. I used the gel with Glycolic Acid and it was bad for my skin, even some pimples appeared.", {"entities": [(80,95, "ReportedTerm")]})]
```
.....
And used to directly pass this in Jupyter notebook to train the model, how do I now convert these set of tuples into a .spacy file for training spacy v3. 

Any help here will be greatly appreciated, have been going through the documentation in circles without any clear instruction set. 

Thanks. 

## Your Environment

- Operating System: Windows10
- Python Version Used:3.8.5
- spaCy Version Used:3
- Environment Information:

**Comment by svlandeg** [comment]
Hi! This is a good question for the discussion forum, so I'll move it there. You may get a notification that your issue was closed/locked, but the thread is still open at the forum and you should be forwarded their automatically :-)

**Comment by adrianeboyd** [comment, marked as answer]
Yes, some basic conversion examples from the v2 `TRAIN_DATA` format have been missing from the docs and projects. The following script will be part of a demo NER project with the full convert/train/package workflow available with [`spacy project`](https://spacy.io/usage/projects) soon.

This script assumes that you've exported `TRAIN_DATA` to a `.json` file with something like `srsly.write_json("train.json", TRAIN_DATA)`:

```python
"""Convert entity annotation from spaCy v2 TRAIN_DATA format to spaCy v3
.spacy format."""
import srsly
import typer
import warnings
from pathlib import Path

import spacy
from spacy.tokens import DocBin


def convert(lang: str, input_path: Path, output_path: Path):
    nlp = spacy.blank(lang)
    docs = []
    for text, annot in srsly.read_json(input_path):
        doc = nlp.make_doc(text)
        ents = []
        for start, end, label in annot["entities"]:
            span = doc.char_span(start, end, label=label)
            if span is None:
                msg = f"Skipping entity [{start}, {end}, {label}] in the following text because the character span '{doc.text[start:end]}' does not align with token boundaries:\n\n{repr(text)}\n"
                warnings.warn(msg)
            else:
                ents.append(span)
        doc.ents = ents
        docs.append(doc)
    db = DocBin(docs=docs)
    db.to_disk(output_path)


if __name__ == "__main__":
    typer.run(convert)
```

**Reply by dayalstrub-cma** [reply]
Thanks, this is great.

Possibly silly question, but would it not be possible to go via `Example`s and would it not make sense for this to be a `Doc` method? What I have in mind is

```python
# sketch implementation
import spacy
from spacy.training import Example
from spacy.tokens import DocBin

nlp = spacy.blank("en")
doc_bin = DocBin()
for text, annotation in data:
    example = Example.from_dict(nlp(text), annotation)
    doc = Doc.from_example(example) # making doc with only gold-standard reference data
    doc_bin.add(doc)
```

This feels like it would connect `Example` and training via script with `DocBin` in .spacy format and using config. Arguably, this won't be an issue when people stop storing as tuples - even though I quite like them.

Note: Unsure whether `docs = []` is more efficient, found `.add` in [documentation](https://spacy.io/usage/saving-loading#docs).

**Reply by dayalstrub-cma** [reply]
After some more digging, saw that you do essentially what I asked about [here](https://github.com/explosion/spaCy/discussions/6435). Is one way better than another?

**Reply by adrianeboyd** [reply]
You're right that it would be better memory-wise to just add the individual docs to the `DocBin`.

Writing the script this way allows for more detailed error messages about misaligned entity spans. The `Example.from_dict` only provides the `W030` warning that some entities are misaligned in a text without the details about which ones.

You don't need anything extra like `Doc.from_example`. You can access the reference doc directly as `example.reference` to add it to the `DocBin`.
